app/src/main/java/com/github/se/eventradar/viewmodel/qrCode/ScanFriendQrViewModel.kt

```diff
@@ -121,16 +211,18 @@ constructor(
     return updateResult is Resource.Success
   }
 
-  fun resetNavigationEvent() {
-    _uiState.value = _uiState.value.copy(action = Action.None)
+  private fun changeAction(action: Action) {
+    _uiState.value = uiState.value.copy(action = action)
+    if (action == Action.NavigateToNextScreen) {
+      navigationActions.navController.navigate(
+          "${Route.PRIVATE_CHAT}/${_uiState.value.decodedResult}") // TODO check correct
+      _uiState.value = _uiState.value.copy(action = Action.None)
```

Comment:
Ok this is something you use quite often in your code (whole project, everyone), this is not thread safe. Here's an example for a better version:

```kotlin
class CounterModel {
    private val _counter = MutableStateFlow(0) // private mutable state flow
    val counter = _counter.asStateFlow() // publicly exposed as read-only state flow

    fun inc() {
        _counter.update { count -> count + 1 } // atomic, safe for concurrent use
    }
}
```

Reply:
FOLLOWUP due to a question regarding this:

---

Nah the atomic thing is more general. It does not concern the navigation, but **view models** in general. 

What we often find in EventRadar is something like `_state.value = state.value.copy(modifiedField = modifiedValue)`. What we need to realise, is that we make a COPY of the current state of `state` and THEN assign this value back to `_state.value`. In other words, it is the same thing as if you were doing

```kotlin
// Note : I am assuming that state.value is mutable in the second line just to make the racing condition point clearer, but the problem is exactly the same if state.value is immutable (because we make a *copy* and *then* assign it back)
val oldValue = state.value
oldValue.modifiedField = modifiedValue
_state.value = oldValue
```

What happens when you have the main thread and a coroutine executing this code? Most of the time, everyhting happens correctly. However, they might sometimes negate each other, e.g.

```kotlin
// thread 1 => makes a copy of `state`
val valueT1 = state.value
valueT1.modifiedField = modifiedValue
_state.value = valueT1 

// thread 2 => makes a copy of `state` w/ the modified state from T1
val valueT2 = state.value // valueT1 in this case
valueT2.modifiedField = modifiedValue
_state.value = valueT2

// Here, _state.value ends up with value T2
```

Then, here's another example where the execution differs a bit

```kotlin
// thread 1 => makes a copy of `state`
val valueT1 = state.value
valueT1.modifiedField = modifiedValue

// thread 2 => makes a copy of `state` WITHOUT T1 changes
val valueT2 = state.value
valueT2.modifiedField = modifiedValue
_state.value = valueT2

// thread1 => finishes the execution (maybe the copy was slow, or the field modified was larger)
_state.value = valueT1

// Here, _state.value ends up with value T1. T
```

Same code, two executions, two different results => that's a problem! Now luckily, you can *very simply* resolve this issue by using `.update` on a mutable state flow, e.g.

```kotlin
_state.update{ it.copy(modifiedField = modifiedValue) }
```

If we take the analogy from before, here's what it would do:

```kotlin
synchronize {
    val oldValue = _state.value
    oldValue.modifiedField = modifiedValue
    _state.value = oldValue
}
```

In other words, `.update` performs the change atomically. Meaning that you won't have any race condition.

Lastly, I want to say that this **is** something that could very well happen in your code since you are sometimes using `LaunchedEffect`. Remember that launch effect *listens* for changes, and then **starts a coroutine** when a change is detected on a particular key